dspace-server-webapp/src/main/java/org/dspace/app/rest/DuplicateDetectionRestController.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest;
+
+import java.sql.SQLException;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.UUID;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.dspace.app.rest.converter.ConverterService;
+import org.dspace.app.rest.model.PotentialDuplicateRest;
+import org.dspace.app.rest.model.hateoas.PotentialDuplicateResource;
+import org.dspace.app.rest.utils.ContextUtil;
+import org.dspace.app.rest.utils.Utils;
+import org.dspace.content.Item;
+import org.dspace.content.service.DuplicateDetectionService;
+import org.dspace.content.service.ItemService;
+import org.dspace.content.virtual.PotentialDuplicate;
+import org.dspace.core.Context;
+import org.dspace.discovery.SearchServiceException;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.rest.webmvc.ResourceNotFoundException;
+import org.springframework.data.web.PagedResourcesAssembler;
+import org.springframework.hateoas.Link;
+import org.springframework.hateoas.PagedModel;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * The controller for the api/duplicates endpoint, which handles requests for finding
+ * potential duplicates
+ *
+ * @author Kim Shepherd
+ */
+@RestController
+@ConditionalOnProperty("duplicate.enable")
+@RequestMapping("/api/" + PotentialDuplicateRest.CATEGORY + "/" + PotentialDuplicateRest.NAME)
```

Comment:
why a new controller is needed? it is a search method so it should be sufficient to create a REST repository following our current architecture. I strongly recommend to get ride of this controller and just introduce the PotentialDuplicateRestRepository

Reply:
@kshepherd : Just to clarify @abollini 's comment more... He's right that this code could be simplified if refactored from `DuplicateDetectionRestController` to `PotentialDuplicateRestRepository` (or whatever you want to name it).  That should not be a large refactor.

Doing that refactor would also fix the feedback @abollini noted on [his REST Contract comment](https://github.com/DSpace/RestContract/pull/252#pullrequestreview-1909161686).

Essentially, you'd just move your existing `DuplicateDectionsRestController.searchPotentialDuplicates()` method into a `PotentialDuplicateRestRepository.searchPotentialDuplicates()` that looks more like this:
```
@PreAuthorize("hasPermission(#uuid, 'ITEM', 'READ')")
@SearchRestMethod(name="byUUID")
public PagedModel<PotentialDuplicateResource> searchPotentialDuplicates(
            @RequestParam(name = "uuid", required = true) UUID uuid, Pageable page, PagedResourcesAssembler assembler) {
```

I don't know if that code is all completely correct... but that "@SearchRestMethod" annotation would result in a subpath like `/search/byUUID?uuid=` under the `/api/submission/duplicates/` path (which would be defined by `PotentialDuplicateRestRepository`).

For many examples of how this works, search for the `@SearchRestMethod` annotation in our codebase...as that is used for almost all subpath `/search` endpoints.   It can only be used in `*RestRepository` class though. 

Apologies, I should have thought of this earlier as it actually should simplify your code a bit...using a *RestRepository also ensures that, by default, you align with other REST endpoint designs.

Reply:
I thought rest repositories were only for addressable objects.
While we do search for an item ID, we are returning stubs which are not themselves an addressable DSO or other db object, just a stub with preview values. I was told to stop using the ItemRestRepository for this very reason in the first review. I will see if it is possible to implement a DuplicateRestRepository, and will switch all my code back to using a searchBy method if that works. I may not have time for this until after the weekend.

Reply:
@kshepherd : I think there's been a misunderstanding?  The problem with using the ItemRestRepository was never about the fact that you _shouldn't use a RestRepository.   Instead, the issue was that the ItemRestRepository class is meant to *always return/represent `Item` resources* (or objects).   In your code, you are **not** returning an Item resource, but are returning something you've named "Duplicate" (`"type": "DUPLICATE"`).   This is a **new type of resource** which is related to an Item, but it is **not** and Item (`"type": "ITEM"`).

So, the original feedback was that you shouldn't use the `ItemRestRepository` simply because your search endpoints don't return `Item` resources.   Instead, you'd need to define your own endpoint (which you've already done) for that new resource...and then create your own RestRepository **or** Controller. 

As a sidenote: In some situations, we've found that Controllers are useful as they allow more control over creating super-custom endpointss.  However, ideally, there's a preference for RestRepository classes because they "force" your endpoints to be well-structured... that "well-structuring" makes it easier to create endpoints that immediately align with our Contract. But it can be limiting for some endpoints that don't fit into this structure.

In all honesty, I should have asked initially *why* you switched to a Controller.  But, this is simply something I wasn't paying attention too as I had several other feature PRs I was attempting to help other developers along with.

Long story short, you can and should use a RestRepository class when it is possible to do so.  There are may non-addressable DSOs / db objects which already have their own RestRepository.  Here's a few: `AuthorizationRestRepository`, `BulkAccessConditionRestRepository`, `ConfigurationRestRepository`, `DiscoveryRestRepository`, ... `VocabularyRestRepository`  (None of these are DSOs or have db objects).  Generally there is a RestRepository for almost every top-level endpoint (with a few exceptions where a Controller is used instead)

I hope that helps, and apologies that I didn't mention/notice this sooner.  This is one of the reasons I always try to get a few reviewers on every PR -- I can miss things too & additional eyes always help.

Reply:
@tdonohue  Sorry, yes I do understand why returning a non-Item from an item repository was bad, I also didn't full explain myself properly in my last comment. I have reverted my controller and tried to adapt these requirements to suit the way duplicates work.

**Summary update**:

I have now implemented a `DuplicateRestRepository` which only implements and registers a SearchRestMethod. It returns Method Not Implemented exceptions for both `findOne` and `findAll`.

To me this is a confusing use of REST repositories -- I thought the ability to at least find an addressable object with `/api/endpointname/{identifier}` was the whole point of them.
Only classes that extend `DSpaceRestRepository<Model, ID>` are able to register `SearchRestMethod` methods so this is what I have had to extend in order to make the required refactory.

**Detailed notes to explain my thought process:**

It is the fact that I cannot use a `@SearchRestMethod` method without extending `DSpaceRestRepository`, which needs addressable objects (am I using this term wrong? "things able to be referenced using IDs" is what I mean), and that doesn't fit the way I have implemented this very simple duplicate detection.

Look at the way the examples you mention are extending `DSpaceRestRepository` -- typed for a REST model and a class of that object's addressable ID -- yes they are not DSOs and not all in the database, but the RestModels are still first-class "things" which can be individually addressed/retrieved with a String ID:

`public class BulkAccessConditionRestRepository extends DSpaceRestRepository<BulkAccessConditionRest, String>`
`public class ConfigurationRestRepository extends DSpaceRestRepository<PropertyRest, String>`
`public class VocabularyRestRepository extends DSpaceRestRepository<VocabularyRest, String>`

There is no string, UUID or integer which can retrieve a single duplicate stub, they are lists of previews generated after comparing an item with everything else in the search index. (but they are not SearchResultsRest lists, unfortunately which again require wrapping directly addressable resources).

So the only exception in that set of examples was `DiscoveryRestRepository`, which only extends `AbstractDSpaceRestRepository`, but this won't allow a `@SearchRestMethod` method to be used -- indeed the actual endpoint mapping still happens on a controller class for this example.

I went through this analysis and thinking when I was moving my code out of `ItemRestRepository` and it is for the reasons above that I thought a controller was the most appropriate solution.
However, I will try to make things work with RestRepository patterns as instructed.

Is the compromise here to "pretend" there is a String that can address a Duplicate object somehow, and then either refuse to implement findOne and findAll, or just mush this implementation later, like treat that string as a search query and return first result for findOne? Or instead use UUID and allow people to get a "duplicate" by item ID to just see what that item *would* look like rendered as a duplicate..?

To avoid confusion I am going for the first option - I will tell DSpaceRestRepository that this extension uses a 'String' ID, but I will throw NOT_IMPLEMENTED exceptions for both `findOne` and `findAll`, and just implement a `SearchRestMethod` as instructed.

Reply:
@kshepherd : Thanks for explaining your point better. I now understand why you were concerned... that said, I like the approach you took.  I think it's OK to have a RestRepository which only implements a `@SearchRestMethod`...and I agree that, in that scenario, the unimplemented methods should be documented to not exist (and throw errors).  Thanks!